src/engine/Service.cpp

```diff
@@ -384,3 +362,44 @@ ProtoResult Service::makeNeutralElementResultForSilentFail() const {
   }
   return {std::move(idTable), resultSortedOn(), LocalVocab{}};
 }
+
+// ____________________________________________________________________________
+void Service::verifyVariables(
+    const nlohmann::json& j,
+    const ad_utility::LazyJsonParser::Generator& gen) const {
+  if (!j["head"].contains("vars") || !j["head"]["vars"].is_array()) {
+    throwErrorWithContext("JSON result does not have the expected structure",
+                          gen.details().first100_);
+  }
+
+  auto vars = j["head"]["vars"].get<std::vector<std::string>>();
+  ad_utility::HashSet<Variable> responseVars;
+  for (const auto& v : vars) {
+    responseVars.emplace("?" + v);
+  }
+  ad_utility::HashSet<Variable> expectedVars(
+      parsedServiceClause_.visibleVariables_.begin(),
+      parsedServiceClause_.visibleVariables_.end());
+
+  if (responseVars != expectedVars) {
+    throwErrorWithContext(
+        absl::StrCat("Header row of JSON result for SERVICE query is \"",
```

Comment:
In this message (as well as the one before) also add a "Probable cause: the remote endpoint sent a JSON response that is not according to the SPARQL standard".